src/lens.c

```diff
@@ -1646,7 +1649,8 @@ static void rtn_rules(struct rtn *rtn, struct lens *l) {
         /* cfg(l1 . l2 ... ln, N) -> N := N1 . N2 ... Nn */
         for (int i=0; i < l->nchildren-1; i++) {
             struct state *s = add_state(prod);
-            RTN_BAIL(rtn);
+            if (s == NULL)
+                goto error;
```

Comment:
IMHO this changes behavior. Can you please describe why it makes sense to replace RTN_BAIL(rtn) with checking s for NULL?

Reply:
Because further in line `add_trans (rtn, s, prod-> end, l-> child);` `s` is being dereferenced inside `add_trans`. I'm not sure if `RTN_BAIL (rtn)` protects against this.

Reply:
Near as I can tell, in the event that memory allocation fails, this is detected within `add_state()` and the error code is stored in `prod->lens->info->error`
According to the comment in src/info.h:

/* There is only one struct error for each Augeas instance */

RTN_BAIL(rtn) checks rtn->info->error and executes `goto error;` on any error.
So the case of `s == NULL` is covered by the pre-existing `RTN_BAIL(rtn)` code.

That's probably beyond the capabilities of the static code analyser to evaluate.
I think the existing code here should stay unchanged.